src/shipit_uplift/shipit_uplift/coverage_by_changeset_impl.py

```diff
@@ -0,0 +1,112 @@
+import whatthepatch
+from shipit_uplift.coverage import coverage_service
+
+
+def get_coverage_builds(changeset):
+    '''
+    This function returns the last successful build before a given
+    changeset and the first successful coverage build after the same
+    changeset.
+    '''
+    r = requests.get('https://hg.mozilla.org/mozilla-central/json-rev/%s' % changeset)
+    push_id = int(r.json()['pushid'])
+
+    # In a span of 15 pushes, we hope we will find successful coverage builds.
+    r = requests.get('https://hg.mozilla.org/mozilla-central/json-pushes?startID=%s&endID=%s' % (push_id - 8, push_id + 7))
+    data = r.json()
+
+    before_changeset = None
+    after_changeset = None
+    pushes = data.items()
+    pushes_before = [(pushid, pushdata) for pushid, pushdata in pushes if int(pushid) < push_id]
+    pushes_after = [(pushid, pushdata) for pushid, pushdata in pushes if int(pushid) >= push_id]
+
+    # Find the last coverage build before the changeset.
+    for pushid, pushdata in sorted(pushes_before, reverse=True):
+        changeset = pushdata['changesets'][-1]
+        if coverage_service.has_coverage(changeset):
+            before_changeset = changeset
+            break
+
+    assert before_changeset is not None
+
+    # Find the first coverage build after the changeset.
+    for pushid, pushdata in sorted(pushes_after):
+        changeset = pushdata['changesets'][-1]
+        if coverage_service.has_coverage(changeset):
+            after_changeset = changeset
+            break
+
+    assert after_changeset is not None
+
+    return (before_changeset, after_changeset)
+
+
+def generate(changeset):
+    previous_build_changeset, next_build_changeset = get_coverage_builds(changeset)
+
+    r = requests.get('https://hg.mozilla.org/mozilla-central/raw-rev/%s' % changeset)
+    patch = r.text
+
+    diffs = []
+
+    for diff in whatthepatch.parse_patch(patch):
+        old_path = diff.header.old_path[2:] if diff.header.old_path.startswith('a/') else diff.header.old_path
+        new_path = diff.header.new_path[2:] if diff.header.new_path.startswith('b/') else diff.header.new_path
+
+        # If the diff doesn't contain any changes, we skip it.
+        if diff.changes is None:
+            continue
+
+        # Retrieve coverage of removed and added lines.
+        new_coverage = coverage_service.get_file_coverage(next_build_changeset, new_path)
+        old_coverage = coverage_service.get_file_coverage(previous_build_changeset, old_path)
+
+        # If we don't have coverage for this file, we skip it.
+        if old_coverage is None and new_coverage is None:
+            continue
+
+        changes = []
+        for old_line, new_line, _ in diff.changes:
+            if old_line is not None and new_line is None:
+                # Removed line.
+                if old_line not in old_coverage or old_coverage[old_line] is None:
+                    coverage = ' '
+                elif old_coverage[old_line] > 0:
```

Comment:
Is the value a count of how many times a line was hit?

Reply:
Exactly. It isn't always precise, especially because we run the tests with optimizations enabled, so I'm not sure we actually want to show it in the UI